File: posixlib/src/main/resources/scala-native/dlfcn.c
#if defined(__unix__) || defined(__unix) || defined(unix) ||                   \
    (defined(__APPLE__) && defined(__MACH__))

#include <dlfcn.h>

int scalanative_rtld_lazy() { return RTLD_LAZY; };

int scalanative_rtld_now() { return RTLD_NOW; };

int scalanative_rtld_global() { return RTLD_GLOBAL; };

int scalanative_rtld_local() { return RTLD_LOCAL; };

#endif // Unix or Mac OS

File: posixlib/src/main/scala/scala/scalanative/posix/dlfcn.scala
package scala.scalanative
package posix

import scala.scalanative.unsafe.*

/** POSIX dlfcn.h for Scala
 *
 *  The Open Group Base Specifications
 *  [[https://pubs.opengroup.org/onlinepubs/9699919799 Issue 7, 2018]] edition.
 */

@link("dl")
@extern
@define("__SCALANATIVE_POSIX_DLFCN")
object dlfcn {

// Symbolic constants

  @name("scalanative_rtld_lazy")
  def RTLD_LAZY: CInt = extern

  @name("scalanative_rtld_now")
  def RTLD_NOW: CInt = extern

  @name("scalanative_rtld_global")
  def RTLD_GLOBAL: CInt = extern

  @name("scalanative_rtld_local")
  def RTLD_LOCAL: CInt = extern

// Methods

  // Convention: A C "void *" is represented in Scala Native as a "CVoidPtr".

  def dlclose(handle: CVoidPtr): Int = extern

  def dlerror(): CString = extern

  def dlopen(filename: CString, flags: Int): CVoidPtr = extern

  def dlsym(handle: CVoidPtr, symbol: CString): CVoidPtr = extern
}
